File: HW2/homework.java
//package hwtrial;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
//package hwtrial;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.*;
/**
 *
 * @author Prerna
 */
public class homework {
    public static void main(String args[]) throws IOException{
        ArrayList<String>lines=new ArrayList<String>();
        BufferedReader in = new BufferedReader(new FileReader("input.txt"));
        String str;
        while((str=in.readLine()) != null){
            lines.add(str);
        }
        
        int n=Integer.parseInt(lines.get(0));
        String atype=lines.get(1);
        String current_player=lines.get(2);
        int max_depth=Integer.parseInt(lines.get(3));
             
        //read the board
        int board[][]=new int[n][n];
        for(int i=4; i< 4+n; i++)
        {
            String line[]=lines.get(i).split(" ");
            for(int j=0; j<n; j++){
                board[i-4][j]=Integer.parseInt(line[j]);
            }
        }
        //read the initial state
        String state="";
        for(int i=4+n; i< 4+n+n; i++){
            String line=lines.get(i);
            state+=line;
        }
        Writer writer = new BufferedWriter(new OutputStreamWriter(
              new FileOutputStream("correct_output.txt"), "utf-8"));
        //create initial object
        BoardState bs=new BoardState(0,state,null, "Start",current_player);
        String ans[];
        if(atype.equals("MINIMAX"))
            ans=bs.minimax(n, max_depth, board, current_player);
        else
            ans=bs.alphabeta(n, max_depth, board, current_player);
        writer.write(ans[1]+"\n");
        for(int i=0; i<n*n; i+=n)
        {    if(i==n*n-1)
                writer.write(ans[0].substring(i,i+n));
            else
                writer.write(ans[0].substring(i,i+n)+"\n");
        }
        writer.close();
    }   
}
class BoardState{
    int depth, x_score, o_score;
    String state, move;
    char now_play, opponent;
    BoardState parent;
    public BoardState(int depth, String state, BoardState parent, String move, String now_play){
        this.depth=depth;
        this.state=state;
        this.parent=parent;
        this.move=move;
        this.now_play=now_play.charAt(0);
        switch(this.now_play){
            case 'X':
                this.opponent='O';
                break;
            case 'O':
                this.opponent='X';
                break;
        }
        this.x_score=0;
        this.o_score=0;
    }
    public int get_score(int board[][], int n, String current_player){
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(this.state.charAt(i*n+j)=='X')
                    this.x_score+=board[i][j];
                else if(this.state.charAt(i*n+j)=='O')
                    this.o_score+=board[i][j];
            }      
        }
        if(current_player.equals("X"))
            return this.x_score-this.o_score;
        else
            return this.o_score-this.x_score;
    }
    public String[] stake(int i, int j, int n){
        String ans[]= new String[2];
        if(i>=0 && j>=0 && i<n && j<n){
            if(this.state.charAt(n*i+j)=='.'){
                StringBuilder sb= new StringBuilder(this.state);
                sb.setCharAt(n*i+j, this.now_play);
                String news=sb.toString();
                char col=(char)(65+j);
                int row=i+1;
                String move=Character.toString(col)+Integer.toString(row)+" "+"Stake";
                ans[0]=news;
                ans[1]=move;                 
            }
        }
        return ans;    
    }
    public ArrayList<String []> can_stake(int n){
        ArrayList<String []> ans=new ArrayList<>();
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                String a[]=this.stake(i, j, n);
                if(a[0]==null && a[1]==null)
                    continue;
                else
                    ans.add(a);
            }
        }   
        return ans;    
    }
    public String raid_neighbors(int newr, int newc, int n, String news){
        StringBuilder sb=new StringBuilder(news);
        int conr, conc;
        //check up
        conr=newr-1;
        conc=newc;
        if(conr>=0 && news.charAt(n*conr+conc)==this.opponent)
            sb.setCharAt(n*conr+conc, this.now_play);
        //check left
        conr=newr;
        conc=newc-1;
        if(conc>=0 && news.charAt(n*conr+conc)==this.opponent)
            sb.setCharAt(n*conr+conc, this.now_play);
        //check right
        conr=newr;
        conc=newc+1;
        if(conc < n &&news.charAt(n*conr+conc)==this.opponent)
            sb.setCharAt(n*conr+conc, this.now_play);
        //check down
        conr=newr+1;
        conc=newc;
        if(conr <n && news.charAt(n*conr+conc)==this.opponent)
            sb.setCharAt(n*conr+conc, this.now_play);
        
        return sb.toString();
    }
    public ArrayList<String []> can_raid(int n){
        ArrayList<String []> ans=new ArrayList<>();
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                //check if current player exists in square
                if(this.state.charAt(i*n+j)==this.now_play){
                    //check if empty squares nearby
                    int newr, newc;
                    //check up
                    newr=i-1;
                    newc=j;
                    if(newr>=0 && this.state.charAt(newr*n+newc)=='.'){
                        String a[]=new String[2];
                        StringBuilder sb=new StringBuilder(this.state);
                        sb.setCharAt(newr*n+newc,this.now_play);
                        char col=(char)(65+newc);
                        int row=newr+1;
                        String move_name=Character.toString(col)+Integer.toString(row)+" "+"Raid";
                        String new_state=this.raid_neighbors(newr, newc, n, sb.toString());
                        a[0]=new_state;
                        a[1]=move_name;
                        //System.out.println(move_name);
                        ans.add(a);  
                        //System.out.println(ans);
                    }
                    //check left
                    newr=i;
                    newc=j-1;
                    if(newc>=0 && this.state.charAt(newr*n+newc)=='.'){
                        String a[]=new String[2];
                        StringBuilder sb=new StringBuilder(this.state);
                        sb.setCharAt(newr*n+newc,this.now_play);
                        char col=(char)(65+newc);
                        int row=newr+1;
                        String move_name=Character.toString(col)+Integer.toString(row)+" "+"Raid";
                        String new_state=this.raid_neighbors(newr, newc, n, sb.toString());
                        a[0]=new_state;
                        a[1]=move_name;
                        //System.out.println(move_name);
                        ans.add(a);      
                        //System.out.println(ans);
                    }
                    //check right
                    newr=i;
                    newc=j+1;
                    if(newc< n && this.state.charAt(newr*n+newc)=='.'){
                        String a[]=new String[2];
                        StringBuilder sb=new StringBuilder(this.state);
                        sb.setCharAt(newr*n+newc,this.now_play);
                        char col=(char)(65+newc);
                        int row=newr+1;
                        String move_name=Character.toString(col)+Integer.toString(row)+" "+"Raid";
                        String new_state=this.raid_neighbors(newr, newc, n, sb.toString());
                        a[0]=new_state;
                        a[1]=move_name;
                        //System.out.println(move_name);
                        ans.add(a);     
                        //System.out.println(ans);
                    }
                    //check down
                    newr=i+1;
                    newc=j;
                    if(newr< n && this.state.charAt(newr*n+newc)=='.'){
                        String a[]=new String[2];
                        StringBuilder sb=new StringBuilder(this.state);
                        sb.setCharAt(newr*n+newc,this.now_play);
                        char col=(char)(65+newc);
                        int row=newr+1;
                        String move_name=Character.toString(col)+Integer.toString(row)+" "+"Raid";
                        String new_state=this.raid_neighbors(newr, newc, n, sb.toString());
                        a[0]=new_state;
                        a[1]=move_name;
                        //System.out.println(move_name);
                        ans.add(a);  
                        //System.out.println(ans);
                    }
                }
            }
        }
        return ans;
    }
    public ArrayList<BoardState> get_children(int n, int limit){
        ArrayList<BoardState> children=new ArrayList<>();
        if(this.depth!= limit){
            ArrayList<String []> nexts=new ArrayList<>();
            nexts.addAll(this.can_stake(n));
            nexts.addAll(this.can_raid(n));
            Iterator<String []> iterator=nexts.iterator();
            while(iterator.hasNext()){
                String pp[]=iterator.next();
                BoardState c= new BoardState(this.depth+1, pp[0],this, pp[1], Character.toString(this.opponent));
                children.add(c);    
                //System.out.println(c.get_score());
            }
        }    
        return children;
    }
    public int find_min(int board[][], int limit, int n, String current_player){
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        if(this.depth== limit || children.isEmpty()){
            int v= this.get_score(board, n, current_player);
            return v;
        }
        else{
            int v= Integer.MAX_VALUE;
            Iterator<BoardState> iterator= children.iterator();
            while(iterator.hasNext()){
                BoardState b= iterator.next();
                v=Math.min(v, b.find_max(board, limit, n, current_player));
            }
            return v;
        }
    }
    public int find_max(int board[][], int limit, int n, String current_player){
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        if(this.depth== limit || children.isEmpty()){
            int v= this.get_score(board, n, current_player);
            return v;
        }
        else{
            int v= -Integer.MAX_VALUE;
            Iterator<BoardState> iterator= children.iterator();
            while(iterator.hasNext()){
                BoardState b= iterator.next();
                v=Math.max(v, b.find_min(board, limit, n, current_player));
            }
            return v;
        }
    }
    public String[] minimax(int n, int limit, int board[][], String current_player){
        String ans[]= new String[2];
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        Iterator<BoardState> iterator= children.iterator();
        int v=-Integer.MAX_VALUE;
        while(iterator.hasNext()){
            BoardState x= iterator.next();
            int i=x.find_min(board, limit, n, current_player);
            if(v<i){
                ans[0]=x.state;
                ans[1]=x.move;
                v=i;                      
            }
        }
        return ans;
    } 
    public String[] alphabeta(int n, int limit, int board[][], String current_player){
        String ans[]= new String[2];
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        Iterator<BoardState> iterator= children.iterator();
        int v=-Integer.MAX_VALUE;
        while(iterator.hasNext()){
            BoardState x= iterator.next();
            int i=x.ab_find_min(board, limit, n, current_player, -Integer.MAX_VALUE, Integer.MAX_VALUE);
            if(v<i){
                ans[0]=x.state;
                ans[1]=x.move;
                v=i;                      
            }
        }
        return ans;
    }
    public int ab_find_max(int board[][], int limit, int n, String current_player, int alpha, int beta){
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        if(this.depth==limit || children.isEmpty()){
            int v= this.get_score(board, n, current_player);
            return v;
        }
        else{
            int v=-Integer.MAX_VALUE;
            Iterator<BoardState> iterator= children.iterator();
            while(iterator.hasNext()){
                BoardState x=iterator.next();
                v=Math.max(v,x.ab_find_min(board, limit, n, current_player, alpha, beta));
                if(v>=beta)
                    return v;
                alpha=Math.max(v,alpha);         
            }
            return v; 
        }    
    }
    public int ab_find_min(int board[][], int limit, int n, String current_player, int alpha, int beta){
        ArrayList<BoardState> children= new ArrayList<>(this.get_children(n, limit));
        if(this.depth==limit || children.isEmpty()){
            int v= this.get_score(board, n, current_player);
            return v;
        }
        else{
            int v=Integer.MAX_VALUE;
            Iterator<BoardState> iterator= children.iterator();
            while(iterator.hasNext()){
                BoardState x=iterator.next();
                v=Math.min(v,x.ab_find_max(board, limit, n, current_player, alpha, beta));
                if(v<=alpha)
                    return v;
                beta=Math.min(v,beta);         
            }
            return v; 
        }       
    }
}

File: HW2/homework.py
import time
class BoardState:
	
	x_score=0
	o_score=0
	my_score=0
	#initialize states with the values of its depth, board, state, maximum dimensions, current player and parent
	def __init__(self,depth,state,now_play,parent,made_move):
		self.depth=depth
		self.state=state
		self.now_play=now_play
		if self.now_play=='X':
			self.opponent='O'
		else:
			self.opponent='X'
		self.parent=parent
		self.childnodes=[]
		self.made_move=made_move
	
	# return the current x and o scores of the board
	def return_scores(self):
		for i in range(0,dim):
	 		for j in range(0,dim):
	 			if self.state[i][j]=='X':
	 				self.x_score+=board[i][j]
	 			elif self.state[i][j]=='O':
	 				self.o_score+=board[i][j]
	 	if next_play=='X':
	 		return self.x_score-self.o_score
	 	else:
	 		return self.o_score-self.x_score

	#if possible, stake the current player at a given location [i,j]
	def stake(self,i,j):
		if (i>=0 and j>=0 and i<dim and j<dim):
			if self.state[i][j]=='.':
				nextstate=[x[:] for x in self.state]
				nextstate[i][j]=self.now_play
				colalph=str(unichr(65+j))
				rowno=str(i+1)
				return [nextstate,colalph+rowno+" Stake"]
	
	#create all stakes of the current player iteratively
	def can_stake(self):
		neighbors=[]
		for i in range(0,dim):
			for j in range(0,dim):
				neighbors.append(self.stake(i,j))
		return neighbors
	
	#once  a raid has been played, check if any neighbors of the new square have been captured 
	#and if yes, change its captor
	def raid_neighbors(self,nextstate,newr,newc):
		#check new neighbours, if they are opponents, capture them
		#check one square up
		if newr-1>=0 and nextstate[newr-1][newc] != self.now_play and nextstate[newr-1][newc]!='.':
			nextstate[newr-1][newc]=self.now_play
		#check one square down
		if newr+1<dim and nextstate[newr+1][newc] != self.now_play and nextstate[newr+1][newc]!='.':
			nextstate[newr+1][newc]=self.now_play
		#check one square left
		if newc-1>=0 and nextstate[newr][newc-1] != self.now_play and nextstate[newr][newc-1] !='.':
			nextstate[newr][newc-1]=self.now_play
		#check one square right
		if newc+1<dim and nextstate[newr][newc+1] != self.now_play and nextstate[newr][newc+1] !='.':
			nextstate[newr][newc+1]=self.now_play
		return nextstate

	#given a current state, check if a raid is possible by the current player and generate the intermediate state
	#before conquering any opponents
	'''
	def can_raid(self):
		neighbors=[]
		for i in range(0,dim):
			for j in range(0,dim):
				if self.state[i][j]==self.now_play:
					
					if i-1>=0 and self.state[i-1][j]=='.':
						nextstate=[x[:] for x in self.state]
						nextstate[i-1][j]=self.now_play
						colalph=str(unichr(65+j))
						rowno=str(i-1+1)
						neighbors.append([self.raid_neighbors(nextstate,i-1,j),colalph+rowno+" Raid"])
						
					if j-1>=0 and self.state[i][j-1]=='.':
						nextstate=[x[:] for x in self.state]
						nextstate[i][j-1]=self.now_play
						colalph=str(unichr(65+j-1))
						rowno=str(i+1)
						neighbors.append([self.raid_neighbors(nextstate,i,j-1),colalph+rowno+" Raid"])
											
					if j+1<dim and self.state[i][j+1]=='.':
						nextstate=[x[:] for x in self.state]
						nextstate[i][j+1]=self.now_play
						colalph=str(unichr(65+j+1))
						rowno=str(i+1)
						neighbors.append([self.raid_neighbors(nextstate,i,j+1),colalph+rowno+" Raid"])
											
					if i+1<dim and self.state[i+1][j]=='.':
						nextstate=[x[:] for x in self.state]
						nextstate[i+1][j]=self.now_play
						colalph=str(unichr(65+j))
						rowno=str(i+1+1)
						neighbors.append([self.raid_neighbors(nextstate,i+1,j),colalph+rowno+" Raid"])
						
		return neighbors
	
	'''
	def can_raid(self):
		neighbors=[]
		for i in range(0,dim):
			for j in range(0,dim):
				if self.state[i][j]=='.':
					
					if (i-1>=0 and self.state[i-1][j]==self.now_play) or (j-1>=0 and self.state[i][j-1]==self.now_play) or (j+1<dim and self.state[i][j+1]==self.now_play) or (i+1<dim and self.state[i+1][j]==self.now_play):
						nextstate=[x[:] for x in self.state]
						nextstate[i][j]=self.now_play
						colalph=str(unichr(65+j))
						rowno=str(i+1)
						neighbors.append([self.raid_neighbors(nextstate,i,j),colalph+rowno+" Raid"])					
		return neighbors
	
	
	#given a current state, check if any children are possible by either staking or raiding	
	def get_children(self):
		if self.depth!=limit:
			children_states=[]
			children=[]
			raid_children=self.can_raid()
			stake_children=self.can_stake()
			children_states=stake_children+raid_children
			for item in children_states:
				if item==None:
					continue
				child=BoardState(self.depth+1,item[0],self.opponent,self,item[1])
				self.childnodes.append(child)
				print child.return_scores(), child.made_move
			return self.childnodes

	#traverse through all children of a given node, and its children till you reach the maximum depth
	
	def find_min(self):
		x=self.get_children()
		if self.depth==limit or not x:
			return self.return_scores()
		else:
			v=float("inf")
			for item in x:
				v=min(v,item.find_max())
			return v

	def find_max(self):
		x=self.get_children()
		if self.depth==limit or not x:
			return self.return_scores()
		else:
			v=-float("inf")
			for item in x:
				v=max(v,item.find_min())
			return v
	
	def minimax(self):
		v=-float('inf')
		consider=None
		for item in self.get_children():
			i=item.find_min()
			if v<i:
				consider=item
				v=i
		return [consider.made_move,consider.state]

	def alpha_beta(self):
		v=-float('inf')
		consider=None
		for item in self.get_children():
			i=item.ab_find_min(-float('inf'),float('inf'))
			if v<i:
				consider=item
				v=i
		return [consider.made_move,consider.state]
		
	def ab_find_max(self,alpha,beta):
		x=self.get_children()
		if self.depth==limit or not x:
			return self.return_scores()
		else:
			v=-float("inf")
			for item in x:
				v=max(v,item.ab_find_min(alpha,beta))
				if v>=beta:
					return v
				alpha=max(alpha,v)
			return v

	def ab_find_min(self,alpha,beta):
		x=self.get_children()
		if self.depth==limit or not x:
			return self.return_scores()
		else:
			v=float("inf")
			for item in x:
				v=min(v,item.ab_find_max(alpha,beta))
				if v<=alpha:
					return v
				beta=min(beta,v)
			return v

#main program with given input
input=open('input.txt','r')
start= time.time()
lines=[]
for line in input:
	lines.append(line)
global dim, next_play, limit, board
dim=int(lines[0].rstrip('\n'))
algo=lines[1].rstrip('\n')
next_play=lines[2].rstrip('\n')
limit=int(lines[3].rstrip('\n'))


board=[]
for i in range(4,4+dim):
	row=[]
	points=lines[i].split()
	for x in points:
		row.append(int(x))
	board.append(row)

state=[]
for i in range(4+dim,4+dim+dim):
	row=[]
	for x in lines[i].strip('\n'):
		row.append(x)
	state.append(row)
initial_state=BoardState(0,state,next_play,None,"start")
of=open('output.txt','w')
if algo=='MINIMAX':
	ans=initial_state.minimax()
	of.write(ans[0]+'\n')
	for item in ans[1]:
		row=""
		for col in item:
			row+=col
		of.write(row+'\n')
else:
	ans=initial_state.alpha_beta()
	of.write(ans[0]+'\n')
	for item in ans[1]:
		row=""
		for col in item:
			row+=col
		of.write(row+'\n')

print time.time()-start
